core/src/main/scala/mongo4cats/client/MongoClient.scala

```diff
@@ -16,55 +16,75 @@
 
 package mongo4cats.client
 
-import cats.effect.{Async, Resource, Sync}
-import cats.syntax.flatMap._
+import cats.~>
+import cats.arrow.FunctionK
+import cats.effect.{Async, Resource}
+import cats.implicits._
+import fs2.Stream
 import com.mongodb.connection.ClusterDescription
 import com.mongodb.reactivestreams.client.{MongoClient => JMongoClient, MongoClients}
-import mongo4cats.bson.Document
 import mongo4cats.database.MongoDatabase
 import mongo4cats.helpers._
+import org.bson.Document
 
 import scala.jdk.CollectionConverters._
 
-abstract class MongoClient[F[_]] {
+trait MongoClient[F[_]] {
   def clusterDescription: ClusterDescription
   def getDatabase(name: String): F[MongoDatabase[F]]
-  def listDatabaseNames: F[Iterable[String]]
-  def listDatabases: F[Iterable[Document]]
-  def listDatabases(session: ClientSession[F]): F[Iterable[Document]]
-  def startSession(options: ClientSessionOptions): F[ClientSession[F]]
-  def startSession: F[ClientSession[F]] = startSession(ClientSessionOptions.apply())
+  def listDatabaseNames: Stream[F, String]
+  def listDatabases(
+      clientSession: Option[ClientSession[F]] = None
+  ): Stream[F, Document]
+  def startSession(
+      options: ClientSessionOptions = ClientSessionOptions.apply()
+  ): F[ClientSession[F]]
+
+  def mapK[G[_]](f: F ~> G): MongoClient[G]
+  def asK[G[_]: Async]: MongoClient[G]
 }
 
-final private class LiveMongoClient[F[_]](
-    private val client: JMongoClient
-)(implicit
-    val F: Async[F]
-) extends MongoClient[F] {
-  def clusterDescription: ClusterDescription = client.getClusterDescription
-
-  def getDatabase(name: String): F[MongoDatabase[F]] =
-    F.delay(client.getDatabase(name)).flatMap(MongoDatabase.make[F])
+object MongoClient {
+  final private case class TransformedMongoClient[F[_]: Async, G[_]](
+      client: JMongoClient,
+      transform: F ~> G
+  ) extends MongoClient[G] {
+    def clusterDescription =
+      client.getClusterDescription
+
+    def getDatabase(name: String) = transform {
+      Async[F].delay {
+        MongoDatabase[F](client.getDatabase(name)).mapK(transform)
```

Comment:
I prefer the `transform(Async[F].delay(client.getDatabase(name)).map(MongoDatabase(_).mapK(transform)))` form to be honest. I don't know how much of an effect this would have in practice, but I prefer the `.delay` to be as small as possible.

It would probably matter here, but just my 2 cents.

Reply:
Ok :+1: